merging tck drivers with new code

xytosis opened the pull request:
I would like to integrate the TCK Java Drivers into this repository, as I think it would be the best organization to have each driver for each implementation live inside their own repos.

The problem I am having is that, for some reason, the channel tests that I wrote that worked for the 2.2 snapshot of the TCP transport does not work with the new TCP code, and the toy server and client in test just doesn't work at all (the client seems to randomly timeout before sending or receiving data from the server).

There is probably something wrong with the way I wrote the code. I think if I can get SimpleServer and Main in the `test` directory working, the driver code should start working too.

yschimke commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/client/JavaTCPClient.java:
```diff
@@ -0,0 +1,119 @@
+
+    public static void main(String[] args) throws MalformedURLException, URISyntaxException {
+        // we pass in our reactive socket here to the test suite
+        String file = "/Users/mjzhu/dev/reactivesocket-java/reactivesocket-tck-drivers/clienttest$.txt";
```
avoid hardcoded paths, a relative path assuming you run the build just inside of reactivesocket-java is probably safer for IDE and gradle use.

xytosis replied:
That makes sense, I was going to allow users to input a command line argument for a path, but I was testing and forgot to change it. Will do it soon.

yschimke commented on reactivesocket-tck-drivers/clienttest$.txt:
```diff
@@ -0,0 +1,111 @@
+!
```
Can you move these under src/(main|test)/resources?

xytosis replied:
Will do, I left them here temporarily to test the driver

yschimke commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/test/PayloadImpl.java:
```diff
@@ -0,0 +1,49 @@
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+
+public class PayloadImpl implements Payload // some JDK shoutout
```
maybe for a separate change, but I believe this move allows us to use io.reactivesocket.util.PayloadImpl

yschimke commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/server/JavaTCPServer.java:
```diff
@@ -0,0 +1,41 @@
+
+    public static void main(String[] args) {
+
+        String file = "/Users/mjzhu/dev/reactivesocket-java/reactivesocket-tck-drivers/servertest$.txt";
```
as above, relative path and hopefully not in module root.

yschimke commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/common/PayloadImpl.java:
```diff
@@ -0,0 +1,72 @@
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+
+public class PayloadImpl implements Payload {
```
as below, I think we can reuse the util.PayloadImpl now.  Could be for a later change.

xytosis replied:
Sounds good, I'll keep this here for now and integrate later.

yschimke commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/common/ParseThread.java:
```diff
@@ -0,0 +1,34 @@
+
+package io.reactivesocket.tckdrivers.common;
+
+public class ParseThread implements Runnable {
```
I'd avoid code that manually creates threads and even implementing runnable.  This class could just be an inline lambda passed to an ExecutorService.

xytosis replied:
I've mostly worked with threads in C before; is ExecutorService the preferred way to do so in Java? I can change this after I get the drivers working again with channel, just so I don't have to potentially debug two things at once.

benjchristensen replied:
I'm going to provide examples with Rx as ParseMarble is begging for the Observable.create(SyncOnSubscribe) generator. That's exactly what it's trying to do ... be a ReactiveStreams Publisher.

yschimke commented:
Is it working as is? Or you still have the issue with channel tests?

Sorry for the driveby review, I'm just back from PTO, getting back up to speed.  I haven't gone through the whole thing and not sure if it previously went through review and discussion in this form in other repo and we just want to do an initial code drop.

xytosis commented:
@yschimke No worries, there's been a few changes since the last time you've looked at the code, but we haven't gone through a rigorous review/discussion yet.

Also, the code is still not working with channel tests after switching to the newest code. Everything works with the 2.2 snapshot, but there seems to be a problem with the ReactiveSocket not waiting and the main thread just exits. In the old code I used ReactiveSocket.startAndWait() but in the new code I'm not sure what to use. If anyone could take a look at the toy server and client in the `test` package and help me get that working, I think I can fix everything else.

yschimke commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/common/TestSubscriber.java:
```diff
@@ -0,0 +1,1113 @@
+    }
+
+    public final boolean awaitAtLeast(long n, long time, TimeUnit unit) throws InterruptedException {
+        numOnNext.await(time, unit);
```
I suggest checking the return values for both these await calls, and making sure the test terminates cleanly if there is some failure causing a missing event.

xytosis replied:
I can do that; I can also look to creating global timeout values as well

yschimke commented:
I suggest also stealing the logging configuration from reactivesocket-cli and making that optionally shown when running the test, so its easier to understand the failure without stepping through a debugger

stevegury commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/client/JavaTCPClient.java:
```diff
@@ -0,0 +1,119 @@
+            URI uri = new URI(target);
+
+            Publisher<ReactiveSocket> socketPublisher = buildConnection(uri);
+            SocketWrapper client = new SocketWrapper();
```
You could replace everything below by:
`Unsafe.blockingSingleWait(socketPublisher, 5, TimeUnit.SECONDS);`

xytosis replied:
Thanks for reviewing my code @stevegury! I used similar and also cleaner line of code from Yuri's CLI project, which has the added benefit of allowing me to add a logger to see the individual frames being sent.

xytosis commented:
@yschimke thanks for all the support you gave me earlier, I was able to find the source of the bug, and the code should be working as is. I will implement the changes you've suggested

yschimke commented:
@xytosis No problem, hopefully this failure you had was a useful test of future implementors experience using the TCK.  

One cherry on top if you have a couple of days free would be to implement the Client Tests as Junit4 Parameterized tests e.g. http://stackoverflow.com/a/359004/1542667  This would be really nice integration in an IDE and allow you to run the tests and just re-run or debug a single failing test case.

But not essential, and specifically not for this review.

yschimke commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/client/JavaClientDriver.java:
```diff
@@ -0,0 +1,517 @@
+/*
+ * Copyright 2016 Facebook, Inc.
```
I believe we agreed to put Netflix as copyright holder for submissions to this repo for simplicity.  cc @benjchristensen

benjchristensen replied:
The Facebook one should be on all these files. Copyrights are whoever submits/edits/touches a file, so there can be multiple at the top of the file.

If someone doesn't care about their copyright, then they can "assign" it to the default Netflix, but we use the Facebook copyright. 

We are however adopting the Apache 2 license since that's what this project uses.

yschimke commented:
Could you add running instructions to reactivesocket-tck-drivers/README.md

Generally I'm ok with this landing sooner rather than later, as in a separate repo we are likely to hit issues continually do to refactoring changes.  But I'd like a thumbsup from @NiteshKant or @stevegury rather than just mine.  They are likely to suffer more of the maintenance burden when changes to the public java API affect the TCK.

NiteshKant commented:
I am on a PTO, so haven't looked at the code. 
Philosophically, what is the reason for having multiple TCK drivers and co-location with respective repos?
If the intent to co-locate to reduce instability with breaking changes, would this continue to exist after we stabilize the implementation?

xytosis commented:
@yschimke Thanks for suggesting the parameterized tests, it would definitely be neat to have and I can look into it once I have documentation set up and when I stabilize the code. I will also start adding documentation today, as well as continue to add documentation to https://github.com/ReactiveSocket/reactivesocket-tck

@NiteshKant Yes, separating the drivers to live with their respective repos would increase maintainability as well as keep the TCK repo polyglot. So ideally we would have wanted to use a stable implementation to implement this driver, but since you are still making changes to the code, every breaking change would have to be reflected within the driver. This would require people to understand how the driver works, but as of now I think the implementation of the driver is mostly straightforward enough for that to not be too big of an issue.

yschimke commented on reactivesocket-tck-drivers/gradlew:
```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env bash
```
you should only need gradlew at the top level

xytosis replied:
I'll get rid of this and try to match the other subprojects as closely as I can.

yschimke commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/client/JavaTCPClient.java:
```diff
@@ -29,26 +29,14 @@
-        cmd.parse(args);
+    public static void run(String realfile, String host, int port, boolean debug2)
+            throws MalformedURLException, URISyntaxException {
+        debug = debug2;
```
rename param and use "this.debug = debug;"

xytosis replied:
It's static so I can't do that, but I'll do something cleaner.

yschimke commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/main/Main.java:
```diff
@@ -0,0 +1,45 @@
+            try {
+                JavaTCPClient.run(file, host, port, debug);
+            } catch (Exception e) {
+                System.out.println("Could not launch client");
```
this seems anti-helpful.  Either catch specific Exceptions where you can provide a simpler and expected error e.g. FileNotFoundException -> "input file not found", or at least print the exception stacktrace.

xytosis replied:
That's a great point, I'll change this so it's more helpful.

benjchristensen replied:
e.printStackTrace() would be preferable to that line.

yschimke commented on reactivesocket-tck-drivers/run:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+java -cp build/libs/reactivesocket-tck-drivers-0.2.2-SNAPSHOT.jar io.reactivesocket.tckdrivers.main.Main $@
```
For scripts the convention is "$@" which keeps the same arguments that were passed in.  I know its counter intuitive, since you expect it to munge them into one.

Try using: https://www.shellcheck.net/ or install with "brew install shellcheck"

xytosis replied:
Sounds good, I'll change this and check out shellcheck.

yschimke commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/client/JavaTCPClient.java:
```diff
@@ -0,0 +1,78 @@
+            throws MalformedURLException, URISyntaxException {
+        debug = debug2;
+        // we pass in our reactive socket here to the test suite
+        String file = "reactivesocket-tck-drivers/src/main/test/resources/clienttest$.txt";
```
I was using a pseudo regex format in my comment, I meant either test or main.

> Can you move these under src/(main|test)/resources?

xytosis replied:
My bad, I'll change this

benjchristensen commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,6 @@
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.12-bin.zip
```
Why is this changed?

xytosis replied:
I'm not too sure... it might have changed when I complied the project as I never opened this file before.

yschimke replied:
Can you revert this unless you know why?  I think I've seen Intellij make this change before.

xytosis replied:
Ok, I can manually revert the change.

stevegury replied:
ditto

benjchristensen commented on reactivesocket-tck-drivers/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,6 @@
+distributionPath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-2.13-bin.zip
```
Why is this file needed here?

xytosis replied:
I'll remove it.

benjchristensen commented on reactivesocket-tck-drivers/run:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
```
I suggest appending the .sh suffix to the file name.

benjchristensen commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/common/EchoSubscription.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+public class EchoSubscription implements Subscription {
```
It would help to have a javadoc explaining what this class is for.

xytosis replied:
This was for setting up echo tests, but I might be able to get rid of this after changing parseMarble

benjchristensen commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/common/TestSubscriber.java:
```diff
@@ -0,0 +1,1115 @@
+        }
+    }
+
+    public final void setEcho(EchoSubscription echosub) {
```
This seems odd to need a one-off concrete implementation here. Why does TestSubscriber need to know about this?

xytosis replied:
You're right, I probably can get rid of this to after changing ParseMarble

benjchristensen commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/common/ParseMarble.java:
```diff
@@ -0,0 +1,165 @@
+/**
+ * This class is rather straight forward for most interactions except for channel.
+ */
+public class ParseMarble {
```
This can probably be turned into an `Iterable` that parses the file. Then you can just do this:

```java
Publisher p = RxReactiveStreams.toPublisher(Observable.from(marbleIterable));
```

Or if something more custom is needed:

```java
    Observable
        .<String> create(SyncOnSubscribe.<Iterator<String>, String> createStateful(
            () -> {
              // parse state ... (don't do this if just an Iterable ... use the above simpler example
              ArrayList<String> list = new ArrayList<String>();
              list.add("one");
              list.add("two");
              list.add("three");
              return list.iterator();
            } ,
            (iter, obs) -> {
              // state == position in iterable

             // this gets called for each onNext to be done so custom parsing could occur

              if (iter.hasNext()) {
                obs.onNext(iter.next());
              } else {
                obs.onCompleted();
              }
              return iter;
            }))
```

NiteshKant commented:
@xytosis IMHO if we do not see the TCK driver code in the main Java repo once the Java repo is stabilized then probably it shouldn't belong here now. 
From stability and overhead of catching up with backward incompatible changes point of view all repos have that issue. Protocol is changing and so are the the implementations.

benjchristensen commented:
I don't understand your comment Nitesh. The point is that the TCK repo should not have all language implementations of TCK drivers. That would make that project nutty with all those languages. The TCK repo contains the TCK test script generator. Then each RS impl also needs the TCK driver in its language. This one is for Java. The C++ repo will end up with its own driver impl. Same with JS. Etc.

xytosis commented:
@NiteshKant I'm not sure I understand what you mean either. I understand that the protocol may change, but it would make sense to have reactivesocket-java specific stuff in this repo, and it would be better and more useful if it were updated here along with the rest of the code instead of sitting in some other place and becomes obsolete because it's forgotten about.

NiteshKant commented:
Nevermind I misunderstood the intent of the TCK driver. Apologies for the confusion.

yschimke commented:
I'm in favour of landing this and iterating with smaller more understandable diffs.  Let's give @NiteshKant and @stevegury today to object, otherwise land and iterate.  Worst case we can temporarily disable the module if it causes build issues.

xytosis commented:
@yschimke, apologies for the late reply, I was out of the office yesterday. However, it doesn't seem like there were any objections, so I will do a quick pass through the code for some last minute changes/fixes and then land it.

stevegury commented on reactivesocket-tck-drivers/build.gradle:
```diff
@@ -0,0 +1,36 @@
+    compile project(':reactivesocket-core')
+    compile project(':reactivesocket-client')
+    compile project(':reactivesocket-transport-tcp')
+    compile project(':reactivesocket-core')
```
duplicate

stevegury commented on reactivesocket-tck-drivers/build.gradle:
```diff
@@ -0,0 +1,36 @@
+    compile project(':reactivesocket-client')
+    compile project(':reactivesocket-transport-tcp')
+    compile project(':reactivesocket-core')
+    compile 'io.netty:netty-all:4.1.3.Final'
```
Remove this and rely on the transitive dependency from reactivesocket-transport-tcp

stevegury commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/client/JavaClientDriver.java:
```diff
@@ -0,0 +1,554 @@
+        for (List<String> t : tests) {
+            TestThread thread = new TestThread(t);
+            thread.start();
+            thread.join();
```
I'm not sure to understand why you use threads here?
(since you start and immediately wait)

xytosis replied:
My idea is that if there are any exceptions, they would be contained within the thread and would allow subsequent tests to run.

stevegury commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/client/JavaClientDriver.java:
```diff
@@ -0,0 +1,554 @@
+            }
+            return Optional.of(hasPassed);
+        }
+        else return Optional.empty();
```
Maybe a better return type would be `boolean`, and use an exception for when `id.size() == 0`

xytosis replied:
That makes sense, I'll change it to that

stevegury commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/client/JavaClientDriver.java:
```diff
@@ -0,0 +1,554 @@
+                ParseChannelThread pct = new ParseChannelThread(pc);
+                pct.start();
+                mypct.set(pct);
+                c.countDown();
```
It seems you don't wait for the thread to finish?
Also, why using a Thread, could you use synchronous code here?

xytosis replied:
I tried both waiting for the thread inside the publisher as well as using synchronous code, but both caused the tests to hang because (my assumption is) that waiting for the `parse()` method to finish inside the publisher delays the actual creation of the publisher until the parse method is finished, which means that while the server is trying to send the client data, the client has not even created it's own publisher and used it to subscribe.

Having the thread run asynchronously from the driver fixes the problem, and I wait for the thread to finish outside of the construction of the publisher.

stevegury replied:
Ok, I understand what you're doing.

stevegury commented on reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/client/JavaClientDriver.java:
```diff
@@ -0,0 +1,554 @@
+    /**
+     * Trivial class that wraps around the ParseChannelThread
+     */
+    private class PCTWrapper {
```
You can use AtommicReference instead

yschimke commented:
OK, merging now.  since its fairly isolated the cost of a blocking issue is small and cheap to resolve.  But its a lot easier to deal with smaller incremental diffs.
